### PR TITLE
Aave Funding Proposal 

### PR DESCRIPTION
**Pull request by efecarranza**
# Notes Compared to Post

AMPL utilization is at 100% and cannot be withdrawn at this time.
BUSD and TUSD will not be withdrawn to keep for offboarding.

# Changelog

Add proposal to fund Aave for the upcoming months.
Add Tests.
Add AIP.

### CODE REVIEW

**Review by sakulstra — CHANGES_REQUESTED** [review]
reminder for me

File: src/20240224_Multi_FundingUpdate/AaveV3Ethereum_FundingUpdate_20240224.sol [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+      address(AaveV3Ethereum.COLLECTOR),
+      0
+    );
+
+    _migrate();
+    _transfer();
+    _swap();
+  }
+
+  function _swap() internal {
```
Comment (sakulstra) [inline src/20240224_Multi_FundingUpdate/AaveV3Ethereum_FundingUpdate_20240224.sol:91]: imo this whole functions is a bit messy:
- transfer & withdraw v3 manually
- transfer & withdraw & swap v2/v3 mix
- swap remaining v3

I think a bit cleaner to review would be:
`_withdrawV2Tokens` -> withdraw v2 tokens to SWAPPER
`_withdrawV3Tokens` -> withdraw v3 tokens to SWAPPER
manually initiate the swaps (as there's no clean way to do `unique` over two arrays.

File: src/20240224_Multi_FundingUpdate/AaveV3Ethereum_FundingUpdate_20240224.sol [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+  uint256 public constant USDC_V3_TO_SWAP = 1_250_000e6;
+  uint256 public constant USDT_V3_TO_SWAP = 1_500_000e6;
+  uint256 public constant USDT_V2_TO_SWAP = 200_000e6;
+
+  uint256 public constant GHO_ALLOWANCE = 2_900_000 ether;
+  uint256 public constant WETH_V3_ALLOWANCE = 600 ether;
+  address public constant ALLOWANCES_WALLET = 0xdeadD8aB03075b7FBA81864202a2f59EE25B312b;
+
+  function execute() external {
+    AaveV3Ethereum.COLLECTOR.approve(
```
[the remark this answers is not in the document]
Comment (efecarranza) [reply src/20240224_Multi_FundingUpdate/AaveV3Ethereum_FundingUpdate_20240224.sol:56]: removed

**Review by rustboyar — APPROVED** [review]

File: src/20240224_Multi_FundingUpdate/AaveV3Ethereum_FundingUpdate_20240224.sol [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+      IERC20(AaveV3EthereumAssets.USDT_UNDERLYING).balanceOf(address(this))
+    );
+
+    AaveV3Ethereum.POOL.deposit(
+      AaveV3EthereumAssets.USDT_UNDERLYING,
+      IERC20(AaveV3EthereumAssets.USDT_UNDERLYING).balanceOf(address(this)),
+      address(AaveV3Ethereum.COLLECTOR),
+      0
+    );
+
```
Comment (sakulstra) [inline src/20240224_Multi_FundingUpdate/AaveV3Ethereum_FundingUpdate_20240224.sol:67]:
```suggestion
```

File: src/20240224_Multi_FundingUpdate/AaveV3Ethereum_FundingUpdate_20240224.sol [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+      AaveV3Ethereum.COLLECTOR.transfer(tokens[i].aToken, address(this), tokens[i].balance);
+
+      AaveV2Ethereum.POOL.withdraw(tokens[i].underlying, type(uint256).max, address(SWAPPER));
+    }
+  }
+
+  /*
+   * @notice Withdraws v3 tokens
+   */
+  function _withdrawV3Tokens() internal {
```
Comment (sakulstra) [inline src/20240224_Multi_FundingUpdate/AaveV3Ethereum_FundingUpdate_20240224.sol:264]:
```suggestion
  function _withdrawV3Tokens() internal {
    uint256 rawUSDTBalance = IERC20(AaveV3EthereumAssets.USDT_UNDERLYING).balanceOf(address(AaveV3Ethereum.COLLECTOR));
    AaveV3Ethereum.COLLECTOR.transfer(
      AaveV3EthereumAssets.USDT_UNDERLYING,
      address(SWAPPER),
      rawUSDTBalance
    );
```

**Review by sakulstra — APPROVED** [review]

**Review by rustboyar — APPROVED** [review]

**Pull request merged**